app/src/main/java/ch/epfl/qedit/backend/permission/AndroidPermManager.java

```diff
@@ -0,0 +1,56 @@
+package ch.epfl.qedit.backend.permission;
+
+import android.content.pm.PackageManager;
+import androidx.core.app.ActivityCompat;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This is the real implementation of the permission manager. Due to how android works, we are
+ * obliged to rely on the permission activity class to send the permission results to this manager.
+ */
+public class AndroidPermManager implements PermissionManager {
+
+    // This is the map of pending requests to the android permission system. nextRequestId
+    // is used to generate unique ids for each request. I assumed that 2^31 values is well
+    // enough for the requestId not to overflow.
+    private final Map<Integer, OnPermissionResult> callbacks;
+    private int nextRequestCode;
+
+    public AndroidPermManager() {
+        this.nextRequestCode = 0;
+        this.callbacks = new HashMap<>();
+    }
+
+    @Override
+    public boolean checkPermission(PermissionActivity activity, String permission) {
+        return ActivityCompat.checkSelfPermission(activity, permission)
+                == PackageManager.PERMISSION_GRANTED;
+    }
+
+    @Override
+    public boolean shouldAskAgain(PermissionActivity activity, String permission) {
+        return ActivityCompat.shouldShowRequestPermissionRationale(activity, permission);
+    }
+
+    @Override
+    public void requestPermissions(
+            PermissionActivity activity, String[] permissions, OnPermissionResult callback) {
+        callbacks.put(nextRequestCode++, callback);
```

Comment:
It is to be sure. But I hope you realize that
`callbacks.put(nextRequestCode++, callback);`

is equivalent to

`callbacks.put(nextRequestCode, callback);`
`nextRequestCode = nextRequestCode + 1;`

Reply:
Yes, that's what I want to do, but thanks for carefully checking that one ! Maybe I can write it in two lines for more clarity if you want.